### ISSUE TITLE
Retain canvas state for high performance drawing

### ISSUE THREAD

**Issue by morious** [flutter/flutter#53805]
Hi,

I would like to be able to continuously draw to a canvas and be able to reflect the canvas state on the screen.

Currently with CustomPainter the paint callback receives a canvas that cannot be set, therefore many developers in the community are relying on passing a points array as state to the custom painter, and then reiterate through the ever growing array redrawing all the points over and over again.

One possible solution: perform the drawing operation into a canvas (member of a widget for ex) and then be able to draw that canvas onto the other canvas in the paint callback of a custom painter by passing it to the custom painter. (HTML5 Canvas allows for drawing one canvas onto another as a reference / Swift allows for drawing CGContext onto another as another reference)

I recently created the following issue: #53804 - which gives a good background to this request.

Thank you for creating this awesome framework!

M

**Comment by chinmaygarde** [comment]
What you are looking for may be a `ui.Picture`. You can create one using a `ui.PictureRecorder`. These are lightweight descriptions of rendering intent. You can perform a draw operation of the picture into the canvas per frame. Pictures are immutable however so you might want to re-construct a picture at set intervals, you can use one picture to form the basis for a new one (though I still would not do this per event, something like a merge once every few operations should work best).

**Comment by benthillerkus** [comment]
Hi, just wanted to add that this is a feature I'd love to see.

This would for example be useful to do a trails effect, where the past content slowly fades out because on each redraw a mostly transparent box is drawn on top of it. Think of the look of retro games or radar for example.

Coming from a web and Processing background, this is something I'd have expected to be able do with canvas. As @chinmaygarde points out, PictureRecorders can be used to achieve such effects, but because you need to redraw everything each frame, it's much more expensive, and needs more setup than just saying "please don't clear this canvas".

**Comment by Sakari369** [comment]
Looking for similar functionality, +1 vote from here.
Would like to draw this tracing effect, where you draw on top of the previous canvas. 

Been looking at this for now some hours and wondering if that is even possible currently, as when you draw to an image with a PictureRecorder, it requires first to create a new Canvas, which creates an empty canvas.

If I want to preserve the contents of that image, or draw on top of the previous image, I dont see an easily intuitive way to do that, maybe I will figure it out but any idea how you could do this kind of tracing effect right now ?

**Comment by Sakari369** [comment]
+1 for support for this.

Spent the day today trying to figure out how to continuosly draw an image on top of an existing canvas, frame by frame, without having to loop each draw call and compose the bitmap from that.

Could not figure out a way to do that at least yet using PictureRecorder, the problem is that the Canvas attached to PictureRecorder is cleared on each frame, and I didn't at least yet find a way to just:

Draw one circle on top of the previous canvas contents, but with a blending operation.

Tested trying to somehow nest two PictureRecorders, one that first captures the circle drawing operation, then returns that as an image. 

This works fine, but now if I try to somehow append this result image with one frame of circle drawn, to a composite image where I would stack these images on top of each other, one per animation frame, it is pretty tricky to see how that can be done, or if it even can be done at all ?

Any points or existing code that might do such a thing ? 

Basically would just like to draw vector shapes on top of each other, frame by frame, without clearing the composition buffer each time I want to append new drawn frames into that.

This is very easy with a canvas, or traditionally just drawing to a buffer which is then blended on top of another buffer, but I can't seem to find any easy way to do this. Do I have to figure out how to write my own shaders, any help appreciated :)

**Comment by Sakari369** [comment]
Okay, found a solution in another issue, did not think of doing it like this, but this piece of code by @pingbird implements this functionality:

https://gist.github.com/pingbird/7674b57025b141872c0c21c1c114ad6d

Thanks @pingbird !

**Comment by Sakari369** [comment]
Would be really happy to see this feature, as PictureRecorder with a background image retained has a bug not freeing the memory, not seeing an easy path to implement this currently.

**Comment by chinmaygarde** [comment]
I understand that having this JustWork™ would be awesome but want to provide some context on why doing this performantly is going to be a challenge. And consequently, the link you provided is going to the sort of compromise you’re going to have to resort to.

When rendering using a canvas to the root surface, Flutter wraps a system compositor provided image and directs graphics commands to it. There is typically a ring buffer of these. While Flutter is rendering into one buffer, the system compositor typically owns the last one since it is in the process of presenting it on screen. So the buffer Flutter is rendering into rendering doesn’t even contain the contents of the last frame. And this is just for the full onscreen image. For parts of the hierarchy, Flutter doesn’t even create a buffer since allocating that would be far too expensive. This is why just adding to the last image is extremely hard. Because, as I said, there is probably not an intermediate last image created for just that part of the Flutter hierarchy, and the full screen buffer Flutter rendering into the last frame is now owned by the system compositor. Now, Flutter could create an image per frame for each sub-hierarchy, but that would be massively expensive and grind almost all frames to halt because of massive memory use.

But, if you know you are going to need the last “frame”s contents, you can save it yourself. This is of course pretty expensive but that expense you can control. This is why is happening in the example you provided. Similarly, the number and nature images you reference in your canvas state is also under you control. When the costs get too high, you flatten it into a new one and start from that checkpoint as the base. It still going to be expensive, but you can control how much. You are doing exactly the kind of expensive things the Flutter engine is desperately trying to avoid. But since you can do it judiciously, you can typically get away with it.

I am not saying that we cannot do better in terms of the number of canvas operations that can be stashed and replayed (and fast). Just hoping to provide some context on why retaining an indefinite number of canvas operations isn’t something that can ever be fast with memory efficiency.

**Comment by cmkweber** [comment]
Also, some gains for that specific sample would be:
- `toImageSync` to avoid the costly flips between CPU -> GPU
- Batching the `drawCircle` commands into a path with [addOval](https://api.flutter.dev/flutter/dart-ui/Path/addOval.html)

**Comment by Sakari369** [comment]
> 
> But, if you know you are going to need the last “frame”s contents, you can save it yourself. This is of course pretty expensive but that expense you can control. This is why is happening in the example you provided. Similarly, the number and nature images you reference in your canvas state is also under you control. When the costs get too high, you flatten it into a new one and start from that checkpoint as the base. It still going to be expensive, but you can control how much. You are doing exactly the kind of expensive things the Flutter engine is desperately trying to avoid. But since you can do it judiciously, you can typically get away with it.
> 
> I am not saying that we cannot do better in terms of the number of canvas operations that can be stashed and replayed (and fast). Just hoping to provide some context on why retaining an indefinite number of canvas operations isn’t something that can ever be fast with memory efficiency.

Thank you for the explanation.
Seems you know a lot about this subject, any chance you could take a look at:

https://github.com/flutter/flutter/issues/144758

And tell me am I doing this wrong ? Or is Flutter not releasing the memory properly for the image retained, and it is an actual bug?

Based on your writing, it should be possible to do this, but currently Flutter just retains the image I want to draw each frame before starting the drawing of normal draw commands, and memory increases all the time and application crashes.

Any help appreciated, thanks!
